hudi-platform-service/hudi-metaserver/hudi-metaserver-client/src/main/java/org/apache/hudi/metaserver/client/HoodieMetaserverClientImp.java

```diff
@@ -0,0 +1,189 @@
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * HoodieMetaserverClientImp based on thrift.
+ */
+public class HoodieMetaserverClientImp implements HoodieMetaserverClient, AutoCloseable, Serializable {
+
+  private static final Logger LOG =  LogManager.getLogger(HoodieMetaserverClientImp.class);
+  private final HoodieMetaserverConfig config;
+  private final int retryLimit;
+  private final long retryDelayMs;
+  private boolean isConnected;
+  private boolean isLocal;
+  private ThriftHoodieMetaserver.Iface client;
+  private TTransport transport;
+
+  public HoodieMetaserverClientImp(HoodieMetaserverConfig config) {
+    this.config = config;
+    this.retryLimit = config.getConnectionRetryLimit();
+    this.retryDelayMs = config.getConnectionRetryDelay() * 1000L;
+    String uri = config.getMetaserverUris();
+    if (isLocalEmbeddedMetaserver(uri)) {
+      try {
+        this.client = (ThriftHoodieMetaserver.Iface) ReflectionUtils.invokeStaticMethod("org.apache.hudi.metaserver.HoodieMetaserver",
+            "getEmbeddedMetaserver", new Object[]{}, new Class[]{});
+      } catch (HoodieException e) {
+        throw new HoodieException("Please check the server uri has ever been set. Empty uri is used for local unit test", e);
+      }
+      this.isConnected = true;
+      this.isLocal = true;
+    } else {
+      URI msUri = URI.create(uri);
+      this.transport = new TSocket(msUri.getHost(), msUri.getPort());
+      this.client = new ThriftHoodieMetaserver.Client(new TBinaryProtocol(transport));
+      try {
+        new RetryHelper<Void, TTransportException>(retryDelayMs, retryLimit, retryDelayMs, TTransportException.class.getName())
+            .tryWith(() -> {
+              transport.open();
+              this.isConnected = true;
+              LOG.info("Connected to meta server: " + msUri);
+              return null;
+            }).start();
+      } catch (TTransportException e) {
+        throw new HoodieException("Fail to connect to the metaserver.", e);
+      }
+    }
+  }
+
+  private boolean isLocalEmbeddedMetaserver(String uri) {
+    return uri == null || uri.trim().isEmpty();
+  }
+
+  @Override
+  public Table getTable(String db, String tb) {
+    return exceptionWrapper(() -> this.client.getTable(db, tb)).get();
+  }
+
+  @Override
+  public void createTable(Table table) {
+    try {
+      this.client.createTable(table);
+    } catch (TException e) {
+      throw new HoodieException(e);
+    }
+  }
+
+  @Override
+  public List<HoodieInstant> listInstants(String db, String tb, int commitNum) {
+    return exceptionWrapper(() -> this.client.listInstants(db, tb, commitNum).stream()
+        .map(EntityConversions::fromTHoodieInstant)
+        .collect(Collectors.toList())).get();
+  }
+
+  @Override
+  public Option<byte[]> getInstantMeta(String db, String tb, HoodieInstant instant) {
```

Comment:
can we rename `InstantMeta` to `InstantMetadata` holistically? and is `InstantMetadata` the equivalence of `CommitMetadata` ? we need to align on the names

Reply:
not really. for compaction.requested, it is compaction plan.